src/04-extensions.js

```diff
@@ -0,0 +1,212 @@
+    for (let i = 0; i < validKeys.length; i++) {
+      if (validKeys[i].toLowerCase() === keyString.toLowerCase()) {
+        keyString = validKeys[i];
+        break;
+      }
+    }
+
+    return keyString;
+  }
+
+  //if key is undefined assign it the string dinosaurId
+  if (key === undefined) key = "dinosaurId";
+  else key = camelCaseKey(key);
+  //edgecase - all dinos share same keys, so check whether key exists in object
+  if (!dinos[0][key] || key === "period") return "please provide a valid key: (name,pronunciation,meaningOfName,diet,lengthInMeters,mya, or info)";
+
+  //initialie an empty object
+  const result = {};
+
+  //iterate through dinosaurs
+  dinos.forEach((dino) => {
+    //if period already exists in object, push dinoID to object at key
+    if (result[dino.period]) result[dino.period].push(dino[key]);
+    //otherwise create key in object and assign it array with current dinoID in it
+    else result[dino.period] = [dino[key]];
+  });
+
+  return result;
+}
+
+const result = organizeDinosaursByPeriod(exampleDinosaurData, "mya");
+// console.log(result);
+
+//--SPACING
+
+/*
+Write a function that takes in an array of strings, where each string is a roomId. Assume that someone is attempting to go from room to 
+room, starting with the room at index 0 and then preceding to the room at index 1, and so on.
+Return true if the path the person is trying to follow is possible. Otherwise, return false.
+*/
+
+const valid = ["aIA6tevTne", "A6QaYdyKra", "L72moIRcrX", "0eNtkY5WoA"];
+const invalid = ["aIA6tevTne", "A6QaYdyKra", "L72moIRcrX", "dpQnu5wgaN"];
+
+function validateMuseumPath(selectedPath, showPath) {
+  //input: array of strings that represents roomIDs and specific order a user wishes to go room to room
+  //output: a boolean value indicating whether the path is possible or not
+
+  //edge-case: empty array
+  if (selectedPath.length === 0) return "array cannot be empty";
+
+  //declare a lastRoom variable and set to null originally
+  let lastRoom = null;
+  //make copy of exampleRoomData
+  remainingRooms = exampleRoomData.slice();
+
+  //make sure they start off at coat check room
+  if (selectedPath[0] !== remainingRooms[2].roomId) {
+    return "path must start at 'aIA6tevTne' a.k.a Coat Check Room";
+  } else {
+    lastRoom = exampleRoomData[2].roomId;
+    selectedPath.slice(1);
+  }
+
+  //declare a variable to chop off remaining room options to prevent repeats, and path variable
+  let indexToChopOffRemainingRooms;
+  let gottenPath = "";
+
+  //iterate through path until we are out of bounds
+  for (let i = 0; i < selectedPath.length; i++) {
+    //if next index is out of bounds return true cause we've hit the end of path successfully
+    if (selectedPath[i + 1] === undefined) break;
+    //declare flag variable which will decide if we proceed so long as next room is valid
+    let nextRoomIsValid = false;
+
+    for (let j = 0; j < remainingRooms.length; j++) {
```

Comment:
I think the only reason to be looping over the overall room data is to:
1. Find the next room's `connectsTo` value.
2. Get the room's name.

I think it might be better to put these into smaller helper functions...